src/transformers/models/llama/modeling_flax_llama.py

```diff
@@ -0,0 +1,701 @@
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+def create_sinusoidal_positions(num_pos, dim):
+    inv_freq = 1.0 / (10000 ** (np.arange(0, dim, 2) / dim))
+    freqs = np.einsum("i , j -> i j", np.arange(num_pos), inv_freq).astype("float32")
+
+    emb = np.concatenate((freqs, freqs), axis=-1)
+    out = np.concatenate((np.sin(emb)[:, None, :], np.cos(emb)[:, None, :]), axis=-1)
+    return jnp.array(out[:, :, :num_pos])  # TODO: don't think slice is needed
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    rotate_half_tensor = jnp.concatenate((-x[..., x.shape[-1] // 2 :], x[..., : x.shape[-1] // 2]), axis=-1)
+    return rotate_half_tensor
+
+
+def apply_rotary_pos_emb(tensor, sincos):
+    sin_pos, cos_pos = sincos
+    return (tensor * cos_pos) + (rotate_half(tensor) * sin_pos)
+
+
+class FlaxLlamaRMSNorm(nn.Module):
+    eps: float = 1e-6
+
+    @nn.compact
+    def __call__(self, hidden_states):
+        input_dtype = hidden_states.dtype
+        variance = jnp.asarray(hidden_states, dtype=jnp.float32)
+        variance = jnp.power(variance, 2)
+        variance = variance.mean(-1, keepdims=True)
+        # use `jax.numpy.sqrt` as `jax.lax.rsqrt` does not match `torch.rsqrt`
+        # hidden_states = hidden_states * jax.lax.rsqrt(variance + self.eps)
+        hidden_states = hidden_states / jnp.sqrt(variance + self.eps)
+
+        weight = self.param("weight", lambda _, shape: jnp.ones(shape), hidden_states.shape[-1])
+
+        return jnp.asarray(weight * hidden_states, dtype=input_dtype)
+
+
+class FlaxLlamaAttention(nn.Module):
+    config: LlamaConfig
+    dtype: jnp.dtype = jnp.float32
+    causal: bool = True
+    is_cross_attention: bool = False
+
+    def setup(self):
+        config = self.config
+        self.embed_dim = config.hidden_size
+        self.num_heads = config.num_attention_heads
+        self.head_dim = self.embed_dim // self.num_heads
+
+        dense = partial(
+            nn.Dense,
+            self.embed_dim,
+            use_bias=False,
+            dtype=self.dtype,
+            kernel_init=jax.nn.initializers.normal(self.config.initializer_range),
+        )
+
+        self.q_proj, self.k_proj, self.v_proj = dense(), dense(), dense()
+        self.o_proj = dense()
+
+        self.causal_mask = make_causal_mask(jnp.ones((1, config.max_position_embeddings), dtype="bool"), dtype="bool")
+        self.embed_positions = create_sinusoidal_positions(config.max_position_embeddings, self.head_dim)
```

Comment:
This style of embedding positions where we don't register them to the state dict works for GPT-J, since both the PyTorch and Flax models compute the embedding positions on the fly

However, for LLaMA, the PyTorch model registers the `inv_freq` to the state dict, and caches the sin/cos positional embeddings on the first forward pass

Now in reality this is fine, since these are not trained model weights, but rather stored constants that can be computed in the `__init__` / `setup` as required. However, we should bear in mind that we'll get weight mis-matches going from PT -> Flax, and **also** from Flax -> PT (i.e. if we load Flax weights in a PyTorch model)

Ideally, we should add these constants to the weight dictionary of the Flax model as well to silence these weight mis-matches -> do you think this is something you could achieve with a small refactor to the embeddings? It's not a big deal if we don't manage, but this would be the ideal situation

Reply:
I think this is doable. We could just make it a variable in the flax model during setup and initialise it in the same way as we would on the fly.

Reply:
@sanchit-gandhi I started looking into this, however the shape of the sinusoidal embeddings is different between the GPT-Neo Flax and Llama PyTorch implementations. Hence, once it is added to the parameter dictionary, it will throw an error during testing as the shapes are different.

We could refactor the Flax version to change the shape, or suppress the warning when there are missing weights for this specific parameter. The latter should be okay as there is no functional difference as far as I can tell. What do you think?

Reply:
@sanchit-gandhi or, third option, like you said:
> It's not a big deal if we don't manage, but this would be the ideal situation

🤔

Reply:
Hey @vvvm23! Did you push code for registering the sinusoidal positional embeddings in the state dict? It would be cool to take a look if you have something to hand, but no worries if not! My preference would be for refactoring the Flax code to match the shape of the PyTorch weights, so that we can load them directly. But if this completely breaks the Flax weights / forward pass with some heavy re-shapes then maybe we can skip. What are the shapes currently in Flax and PyTorch?

Reply:
Haven't yet, was waiting confirmation either way 😅 I can give it a shot tomorrow. I think the main difference is just the order of dimensions, just might affects masking and caching behaviour too but hopefully not in a tricky way. Will update you tomorrow.

Reply:
@sanchit-gandhi the solution to this ended up being quite simple. I had misudnerstand what `inv_freq` did.

In the PyTorch code:
- `inv_freq` is a tensor of size `head_dim // 2` containing the reciprocal of increasing frequencies.
- This tensor is cached as part of the state dictionary.
- This tensor can be used to derive `cos_cached` and `sin_cached` which are also registered as buffers but with `persistent=False` so they do not become part of the saved state dictionary.

In my Flax code:
- `inv_freq` is calculated and immediately used to derive `cos_cached` and `sin_cached`.
- `inv_freq` is immediately discarded and overridden with the same name to the concatenation of `cos_cached` and `sin_cached`

So, I had mistakedly given `inv_freq` the wrong variable name – it is more like `sincos`.

So to fix, I just create the tensor `inv_freq` as in the PyTorch code. This is unused in the Flax, but will be used when loading back to a PyTorch model. This stops the warning.

Sorry for my mistake 😓